### PR TITLE
Multiple messages with the same old name

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
#  Description

Permit to have multiple messages with the same old name. There are optimisations done that make Pylint a little faster.

## Type of Changes
|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

## Related Issue

This is it, after this PR the real review for #1164 can start.

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
I made some performance review (This is why I removed the MessageId class). Regarding performance, I think this is now making things better but it's not very clear cut:

# Before

```
time python3 -m cProfile -s tottime pylint/__main__.py pylint/ > result
python3 -m cProfile -s tottime pylint/__main__.py pylint/ > result
68,30s user 0,46s system 99% cpu 1:08,81 total
cat result|grep message_|cut -c 14-18|paste -sd+ | bc                  
.453
```
Longest message function before : 

```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
40359    0.138    0.000    0.369    0.000 message_handler_mix_in.py:175(is_message_enabled)
51443    0.122    0.000    0.140    0.000 message_definition_store.py:152(get_message_definitions)
40359    0.070    0.000    0.091    0.000 message_handler_mix_in.py:197(is_one_message_enabled)
40358    0.028    0.000    0.028    0.000 message_handler_mix_in.py:186(<listcomp>)
9800    0.013    0.000    0.041    0.000 message_handler_mix_in.py:153(_message_symbol)
```


# After
```
time python3 -m cProfile -s tottime pylint/__main__.py pylint/ > result
python3 -m cProfile -s tottime pylint/__main__.py pylint/ > result
65,46s user 0,38s system 99% cpu 1:05,92 total
$ cat result|grep message_|cut -c 14-18|paste -sd+ | bc                  
.573
```
Longest messages functions after :
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 40396    0.166    0.000    0.542    0.000 message_handler_mix_in.py:173(is_message_enabled)
    51484    0.130    0.000    0.187    0.000 message_id_store.py:108(get_active_msgids)
    51484    0.097    0.000    0.325    0.000 message_definition_store.py:52(get_message_definitions)
    40396    0.066    0.000    0.085    0.000 message_handler_mix_in.py:195(is_one_message_enabled)
    51477    0.040    0.000    0.040    0.000 message_definition_store.py:60(<listcomp>)
```

So it seems that overall the analysis is faster, but the time sepcifically for the part we changed is a little longer. We could probably merge MessageIdStore and MessageDefinitionStore to have less function call (51484 call to get_active_msgids are added) but I think it would affect maintenability a little.

**Comment by Pierre-Sassoulas** [comment]
@PCManticore this one is ready for review, it should be a piece of cake after #2992 ;)

File: pylint/message/message_id_store.py [hunk 1]
```diff
@@ -120,4 +120,9 @@ def get_active_msgids(self, msgid_or_symbol: str) -> List[str]:
                 msgid_or_symbol=msgid_or_symbol
             )
             raise UnknownMessageError(error_msg)
-        return [msgid]
+        # logging.debug(
+        #    "Return for {} and msgid {} is {}".format(
+        #        msgid_or_symbol, msgid, self.__old_names.get(msgid, [msgid])
+        #    )
+        # )
+        return self.__old_names.get(msgid, [msgid])
```
Comment (Pierre-Sassoulas) [inline pylint/message/message_id_store.py]: This line is the one that makes multiple messages with the same old name possible, everything else is tests or optimizations.

**Comment by Pierre-Sassoulas** [comment]
Hi, @PCManticore, I won't be super available from mid-September to the end of October, now would be the best time to review this and #3021 :)

**Comment by PCManticore** [comment]
Awesome stuff @Pierre-Sassoulas ! 👍Sorry for the delay in getting to this, I was out as well.

**Pull request merged**